src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,37 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::*;
+use rustc_data_structures::fx::FxHashSet;
+use crate::transform::{MirPass, MirSource};
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(&self, _tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let mut unreachable_blocks = FxHashSet::default();
+
+        for (bb, bb_data) in traversal::postorder(body) {
+            if let Some(terminator) = &bb_data.terminator {
+                if let TerminatorKind::Unreachable = terminator.kind {
+                    unreachable_blocks.insert(bb.clone());
+                } else {
+                    let mut bb_successors = terminator.successors().peekable();
+
+                    if bb_successors.peek().is_some() {
+                        if bb_successors.all(|bb| unreachable_blocks.contains(bb)) {
+                            unreachable_blocks.insert(bb.clone());
+                        }
+                    }
+                }
+            }
+        }
+
+        for block in unreachable_blocks {
+            body.basic_blocks_mut()[block]
+                .terminator.as_mut().unwrap().kind = TerminatorKind::Unreachable;
+        }
```

Comment:
Suppose we have:
```rust
bb1: {
    goto -> bb2;
}

bb2: {
     unreachable;
}
```
At this point, you would have changed the graph to:

```rust
bb1: {
    unreachable;
}

bb2: { // Nothing refers to this BB anymore.
     unreachable;
}
```
I believe we have checks that would not allow this (indeed I hit those checks).
To remedy this, I believe you have to call `simplify::remove_dead_blocks` (but only if you actually changed something).

Reply:
Could you please tell me which checks these are?

Reply:
The dead blocks get removed by one of the following MIR stages anyway. Is it worthwhile doing the drop here? I think the question that's not been asked here is: 

What would be the best position in the list of MIR transforms for this transform to be in. Should it be one of the first transforms to do (or are there some transforms that might create unreachables that should come before this transform)?

Reply:
> The dead blocks get removed by one of the following MIR stages anyway.

That depends on whether those passes want to call `is_dominated_by`. ;)

I intentionally messed things up now in my branch and got:
```
   9: rustc_data_structures::graph::dominators::Dominators<Node>::is_dominated_by
  10: rustc_codegen_ssa::mir::analyze::LocalAnalyzer<Bx>::process_place
  11: rustc_codegen_ssa::mir::analyze::non_ssa_locals
  12: rustc_codegen_ssa::mir::codegen_mir
  13: rustc_codegen_ssa::base::codegen_instance
  14: <rustc::mir::mono::MonoItem as rustc_codegen_ssa::mono_item::MonoItemExt>::define
  15: rustc_codegen_llvm::base::compile_codegen_unit::module_codegen                                     
  16: rustc::dep_graph::graph::DepGraph::with_task
  17: rustc_codegen_llvm::base::compile_codegen_unit
  18: rustc_codegen_ssa::base::codegen_crate
  19: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  20: rustc_interface::passes::start_codegen::{{closure}}
  21: rustc::util::common::time
  22: rustc_interface::passes::start_codegen
  23: rustc::ty::context::tls::enter_global
  24: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  25: rustc_interface::passes::create_global_ctxt::{{closure}}
  26: rustc_interface::passes::BoxedGlobalCtxt::enter
  27: rustc_interface::queries::Query<T>::compute
  28: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen

  29: rustc_interface::interface::run_compiler_in_existing_thread_pool
  30: std::thread::local::LocalKey<T>::with
  31: scoped_tls::ScopedKey<T>::set
  32: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```